io-engine/src/lvs/lvs_lvol.rs

```diff
@@ -442,17 +440,182 @@ impl Lvol {
-                .to_result(|error| Error::SnapshotCreate {
-                    source: Errno::from_i32(error),
-                    msg: c_snapshot_name.into_string().unwrap(),
-                }),
+        let (error, lvol) =
+            receiver.await.expect("Snapshot done callback disappeared");
+        match error {
+            0 => Ok(lvol),
+            _ => Err(Error::SnapshotCreate {
+                source: Errno::from_i32(error),
+                msg: c_snapshot_name.into_string().unwrap(),
+            }),
         }
     }
+    async fn do_create_snapshot_remote(
+        &self,
+        snap_param: SnapshotParams,
+        done_cb: unsafe extern "C" fn(*mut c_void, *mut spdk_lvol, i32),
+        done_cb_arg: *mut ::std::os::raw::c_void,
+    ) -> Result<(), Error> {
+        let mut attr_descrs: [spdk_xattr_descriptor; SnapshotXattrs::COUNT] =
+            [spdk_xattr_descriptor::default(); SnapshotXattrs::COUNT];
+
+        // Vector to keep allocated CStrings before snapshot  creation
+        // is complete to guarantee validity of attribute buffers
+        // stored inside CStrings.
+        let mut cstrs: Vec<CString> = Vec::new();
+
+        self.prepare_snapshot_xattrs(
+            &mut attr_descrs,
+            snap_param.clone(),
+            &mut cstrs,
+        )?;
+
+        let c_snapshot_name = snap_param.name().unwrap().into_cstring();
+
+        // No need to flush blob's buffers explicitly as SPDK always
+        // synchronizes blob when taking a snapshot.
+        unsafe {
+            vbdev_lvol_create_snapshot_ext(
+                self.as_inner_ptr(),
+                c_snapshot_name.as_ptr(),
+                attr_descrs.as_mut_ptr(),
+                SnapshotXattrs::COUNT as u32,
+                Some(done_cb),
+                done_cb_arg,
+            )
+        };
+        Ok(())
+    }
+    /// Common API to get the xattr from blob.
+    pub fn get_blob_xattr(
+        snapshot: &Lvol,
+        attr: &SnapshotXattrs,
+    ) -> Option<String> {
+        let mut val = std::ptr::null();
+        let mut size: u64 = 0;
+        let attr_id = attr.name().to_string().into_cstring();
+        let curr_attr_val = unsafe {
+            let blob = snapshot.bs_iter_first();
+            let r = spdk_blob_get_xattr_value(
+                blob,
+                attr_id.as_ptr(),
+                &mut val as *mut *const c_void,
+                &mut size as *mut u64,
+            );
+            // mark snapshot invalid, if any attribute not found
+            if r != 0 {
+                warn!(
+                    "Snapshot attribute {:?} not found, snapshot{:?}",
+                    attr_id, snapshot
+                );
+                return None;
+            }
+            String::from_raw_parts(val as *mut u8, size as usize, size as usize)
+        };
+        Some(curr_attr_val)
+    }
+    /// Common API to set SnapshotDescriptor for ListReplicaSnapshot.
+    pub fn set_snapshot_descriptor(
```

Comment:
name is a bit misleading, seems we're setting something
```suggestion
    pub fn snapshot_descriptor(
```